otwcode/otwarchive
# AO3-5878 Add error to invalid prompt update

## Issue
https://otwarchive.atlassian.net/browse/AO3-5878

## Purpose
After signing up for a challenge or exchange, users can edit their requests and offers in the collection by clicking on "My Sign-up"/"My Prompts" in the sidebar. Each request/offer/prompt has an "Edit Prompt"-button, which leads to a form where the specific prompt can be edited.

Currently, if the user changes the prompt so it becomes invalid (i.e. doesn't add the required number of character tags, or similar), the page simply reloads without returning an error message or applying the changes.

The committed change adds a detailed error message to a failed prompt update in the edit prompt view using the validation helper. It also adds a check for stray "^" signs into the error formatting of the validation helper to prevent prompt errors being rendered with a "^"  at the beginning.

## Testing Instructions

**Bug fix test:**

1. Create challenge: Gift Exchange or Prompt Meme
2. Define required challenge attributes: e.g. 2 required characters, or 2 required fandoms ...
3. Sign up for the challenge using correct attributes
4. Visit the collection home page ( .../collections/[your test collection]) and click on "My Sign-up" (Gift Exchange) or "My Prompts" (Prompt Meme) in the sidebar
5. Edit any prompt by using the "Edit Prompt" button
6. Change attributes so they don't comply with the challenge requirements anymore and click "Update"

Expected outcome: page reloads showing a detailed, correctly formatted error message telling you which prompt attributes don't comply with the challenge settings, similar to what happens when you try to edit a prompt on the "Edit Sign-Up" page.

**Regression tests for the validation helper:**

These should not be necessary in all places it's being used (61 views), since the change only adds a search for "^" and removes all found instances. Instances should only be found if the "^" that marks the beginning of the error message has not already been removed (as was the case with the rendering of the error message in the prompt edit view), in which case removal is desired.

For a selective regression test, proceed as described above. In Step 5, use the "Edit Sign-up" instead of the "Edit Prompt" button. Expected outcome: page reloads showing detailed and correctly formatted error messages.

outdated:

> The committed change adds an error flash to the prompts controller to be shown if a user tries to update invalid prompt changes.
> 
> The Jira issue asks for a similarly detailed message as is shown when a user tries to submit an invalid change with the sign-up form ("Edit Sign-up"-button on the "My Sign-up" page). Since the sign-up form uses a different controller than the prompts form, though, this would be quite complicated to implement. I decided to go with a simple error flash which mentions that more details can be gained from using the "Edit Sign-up" function:
> 
> ![image](https://user-images.githubusercontent.com/23420872/78598299-5d393080-784f-11ea-9ad1-8a9f52c65e62.png)
> 
> I also added the ts()-element to the successful update prompt for consistency.
> 
> ## Testing Instructions
> 
> 1. Create challenge: Gift Exchange or Prompt Meme
> 2. Define required challenge attributes: e.g. 2 required characters, or 2 required fandoms ...
> 3. Sign up for the challenge using correct attributes
> 4. Visit the collection home page ( .../collections/[your test collection]) and click on "My Sign-up" (Gift Exchange) or "My Prompts" (Prompt Meme) in the sidebar
> 5. Edit any prompt by using the "Edit prompt" button
> 6. Change attributes so they don't comply with the challenge requirements anymore and click "Update"
> 
> Expected outcome: error message (see above) is shown.

## References

n/a

## Credit

*What name do you want us to use to credit your work in the [Archive of Our Own's Release Notes](https://archiveofourown.org/admin_posts?tag=1)?*

teyla

*What pronouns do you prefer (she/her, he/him, zie/hir etc)?*

she or they
